ce/api/health.py

```diff
@@ -0,0 +1,113 @@
+                            'tasmax_sClim_BNU-ESM_historical_r1i1p1_19650101-19701230': '2020-03-16T17:47:52Z'
+                            'tasmax_mClim_BNU-ESM_historical_r1i1p1_19650101-19701230': '2020-03-16T17:47:54Z'
+                            ...
+                        },
+                        'outdated': {
+                            'pr_aClim_BNU-ESM_historical_r1i1p1_19650101-19701230': '2020-03-16T17:49:03Z'
+                            'pr_sClim_BNU-ESM_historical_r1i1p1_19650101-19701230': '2020-03-16T17:49:06Z'
+                            ...
+                        },
+                        'removed': {
+                            ...
+                        }
+                    }
+                    "vancouver_island.csv": {
+                        ...
+                    }
+                    ...
+                }
+            }
+    '''
+
+    # check to see if any of the stored queries are out of date.
+    # first get metadata on current datasets from the database:
+    current_metadata = multimeta(sesh, ensemble_name=ensemble_name)
+    date_format = '%Y-%m-%dT%H:%M:%SZ'
+    summarize_list = list_files.lower() != 'true'
+
+    # open each stored query file and make sure all data is up to date
+    regional_checks = {}
+    files_checked = {}
+    region_dir = os.getenv('REGION_DATA_DIRECTORY').rstrip("/")
+    for region in os.listdir(region_dir):
+        with open("{}/{}".format(region_dir,
+                                 region), "r") as region_query_file:
+            region_queries = DictReader(region_query_file)
+            region_status = {
+                "current": {},
+                "outdated": {},
+                "removed": {}
+                }
+            for row in region_queries:
+                uid = row['unique_id']
+                modtime = row['modtime']
+                if uid not in files_checked or files_checked[uid] != modtime:
+                    if uid not in current_metadata:
+                        # this stored query is from a
+                        # dataset that has since been deleted!
+                        region_status["removed"][uid] = modtime
+                    elif datetime.strftime(current_metadata[uid]["modtime"],
+                                           date_format) != modtime:
+                        # this stored query is from a
+                        # dataset that has since been updated
+                        region_status["outdated"][uid] = modtime
+                    else:
+                        # this stored data is still accurate
+                        region_status["current"][uid] = modtime
+                    files_checked[uid] = modtime
+
+        regional_checks[region] = {k:len(v) for (k,v) in region_status.items()} if summarize_list else region_status
+
+    return {
+        "stored_regional_queries": regional_checks
+        }
```

Comment:
My original thoughts were that we'd eventually add a `cache` key to the `/health` endpoint in addition to `stored_regional_queries`, but given our experiments in how long it takes to resolve a stored queries check, we definitely want to check cache much more frequently than stored query status. So it would make sense to separate them; something like `/health/files` and `/health/cache` would be easier to configure monitoring for.

I was hoping we'd be able to integrate this check into some kind of monitoring tool, and the tool would determine the format of the response, but at least at the moment we don't have a good option for that. So I didn't put a lot of thought into the response format, assuming it would be changed to match some monitoring tool's standards. As we aren't immediately don't that, I appreciate your suggestions, and I'll update it.

----------

`percentileanomaly`'s format is intended to mimic `timeseries` as far as possible, since it acts as a stand-in for `timeseries` in cases where multiple calls to `timeseries` are too slow. I'm happy to change it if moving toward a more RESTful API is more important than congruence between existing and new API endpoints and you have time to update the frontend.

Reply:
:+1: to the plan of extending the health endpoints with new URL paths rather than keys in the result. I am glad to hear my guess about monitoring configuration confirmed.

Regarding `percentileanomaly`, on a closer look my opinion is that the outer level of the response is correct (i.e., a dict), since it is a specific item and not at that level a list of items. (We have no explicit list endpoint for these objects; or we do but it is a matter of unravelling the metadata response to find out what the variables, climatology years, etc. are.) The only things that *might* benefit from a change of format would be the actual values; for example:

```
{
  "anomaly": [
    {
      "timescale": "yearly",
      "index": 16,
      "date": "2055-07-02 00:00:00",
      "values": [ 2.3264379226208, 3.2458067672088 ]
    },
    {
      "timescale": "seasonal",
      "index": 15,
      "date": "2055-01-16 00:00:00",
      "values": [ 2.7983318484797, 3.3868735586218 ]
    },
    ...
  ]
}
```

This would be considerably easier to filter for the item of interest, but we've already got code committed to filtering the other format. OTOH, a *lot* of that code would just go away, which is kind of the point. How about we make this a separate issue, and coordinate the work on the frontend when we both have time for it, which I do not at the moment.

Reply:
[Issue](https://github.com/pacificclimate/climate-explorer-backend/issues/133) added.